jujugui/static/gui/src/app/state/state.js

```diff
@@ -0,0 +1,323 @@
+      should be the full base path that the GUI is served from.
+      @property baseURL
+      @type {String}
+    */
+    if (!cfg.baseURL) {
+      throw 'baseURL must be provided.';
+    }
+    this.baseURL = cfg.baseURL;
+    /**
+      The list of dispatchers for the various components from the url.
+      @type {Object}
+    */
+    this.dispatchers = cfg.dispatchers;
+
+    if (cfg.seriesList && !Array.isArray(cfg.seriesList)) {
+      throw 'Series list must be an Array.';
+    }
+    /**
+      The list of possible distro 'series' for the store paths. ex)
+      trusty, precise, xenial. The GUI has a utils method `getSeriesList` that
+      is the central storage location for the series information. This Object
+      should be converted to an array and passed to this state when being
+      instantiated.
+      @type {Array}
+    */
+    this.seriesList = (function() {
+      if (!cfg.seriesList) {
+        console.log('No series list provided, using default list');
+        return ['bundle', 'precise', 'trusty', 'xenial'];
+      } else {
+        return cfg.seriesList.concat(['bundle']);
+      }
+    }());
+  }
+
+  /**
+    Takes a complete URL, strips off the supplied baseURL and extra slashes.
+    @param {String} url - The URL to sanitize.
+    @return {String} The sanitized path.
+  */
+  _getCleanPath(url) {
+    return url
+      // Strip the baseURL before parsing the sections.
+      .replace(this.baseURL, '')
+      // Strip the leading and trailing slashes off the url.
+      .replace(/^\/*/, '').replace(/\/*$/, '');
+  }
+
+  /**
+    Splits the URL up and generates a state object.
+    @param {String} url - The URL to turn into state.
+    @return {Object} The generated state object.
+  */
+  buildState(url) {
+    let error = null;
+    let state = {};
+    let parts = this._getCleanPath(url).split('/');
+    state = this._parseRoot(parts, state);
+    // If we have root paths in the URL then we can ignore everything else.
+    if (state.root) {
+      // If there is anything after this then it's an invalid url.
+      if (parts.length > 1) {
+        error = 'invalid root path.';
+      }
+      return {error, state};
+    }
+    // The order of the PATH_DELIMETERS is important so we can assume the
+    // order for easy parsing of the path.
+    if (parts[0] === PATH_DELIMETERS.get('search')) {
+      state = this._parseSearch(parts.splice(1), state);
+      // If we have a search path in the URL then we can ignore everything else.
+      return {error, state};
+    }
+    // Working backwards to split up the URL.
+    const guiIndex = parts.indexOf(PATH_DELIMETERS.get('gui'));
+    if (guiIndex > -1) {
+      ({state, error} = this._parseGUI(parts.splice(guiIndex), state));
+      if (error !== null) {
+        return {error, state};
+      }
+    }
+    // Extract out the user sections.
+    if (parts.includes(PATH_DELIMETERS.get('user'))) {
+      ({state, parts, error} = this._parseUser(parts, state));
```

Comment:
This may be the first new JS feature that I really like, except for the need of surrounding parens.